File: src/utils/monthlyTarget.ts
import type { UserData } from "../types";

export const MAX_MONTHLY_TARGET_MINUTES = 31 * 24 * 60;

export function parseMonthlyTargetInput(value: string): number | undefined {
  const match = /^(\d{1,3}):([0-5]?\d)$/.exec(value.trim());
  if (!match) return undefined;
  return normalizeMonthlyTargetMinutes(Number(match[1]) * 60 + Number(match[2]));
}

export function formatMonthlyTargetInput(minutes: number | undefined): string {
  const normalized = normalizeMonthlyTargetMinutes(minutes);
  if (normalized === undefined) return "";
  return `${Math.floor(normalized / 60)}:${String(normalized % 60).padStart(2, "0")}`;
}

export function normalizeMonthlyTargetMinutes(value: unknown): number | undefined {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value <= 0 ||
    value > MAX_MONTHLY_TARGET_MINUTES
  ) {
    return undefined;
  }
  return value;
}

/** Erhaelt unbekannte Legacy-Felder, entfernt aber ein beschaedigtes Monatsziel. */
export function normalizeUserDataMonthlyTarget<T extends Record<string, unknown>>(userData: T): T {
  const target = normalizeMonthlyTargetMinutes(userData.monthlyTargetMinutes);
  if (target !== undefined) return userData;
  if (!("monthlyTargetMinutes" in userData)) return userData;
  const { monthlyTargetMinutes: _invalid, ...rest } = userData;
  return rest as T;
}

export interface MonthlyTargetProgress {
  targetMinutes: number;
  actualMinutes: number;
  differenceMinutes: number;
  remainingMinutes: number;
  exceededMinutes: number;
  progressPercent: number;
}

export function getValidMonthlyTargetMinutes(userData?: UserData | null): number | null {
  if (!userData?.simpleMode) return null;
  return normalizeMonthlyTargetMinutes(userData.monthlyTargetMinutes) ?? null;
}

export function calculateMonthlyTargetProgress(
  actualMinutes: number,
  userData?: UserData | null,
): MonthlyTargetProgress | null {
  const targetMinutes = getValidMonthlyTargetMinutes(userData);
  if (targetMinutes === null) return null;

  const safeActual = Number.isFinite(actualMinutes) ? Math.max(0, Math.trunc(actualMinutes)) : 0;
  const differenceMinutes = safeActual - targetMinutes;
  return {
    targetMinutes,
    actualMinutes: safeActual,
    differenceMinutes,
    remainingMinutes: Math.max(0, -differenceMinutes),
    exceededMinutes: Math.max(0, differenceMinutes),
    progressPercent: Math.min(100, (safeActual / targetMinutes) * 100),
  };
}

File: native/core/src/main/kotlin/com/estundnzettl/core/calc/MonthlyTarget.kt
package com.estundnzettl.core.calc

import com.estundnzettl.core.model.UserData

const val MAX_MONTHLY_TARGET_MINUTES: Int = 31 * 24 * 60

fun normalizeMonthlyTargetMinutes(value: Int?): Int? =
    value?.takeIf { it in 1..MAX_MONTHLY_TARGET_MINUTES }

fun parseMonthlyTargetInput(value: String): Int? {
    val match = Regex("^(\\d{1,3}):([0-5]\\d)$").matchEntire(value.trim()) ?: return null
    return normalizeMonthlyTargetMinutes(match.groupValues[1].toInt() * 60 + match.groupValues[2].toInt())
}

fun formatMonthlyTargetInput(minutes: Int?): String {
    val value = normalizeMonthlyTargetMinutes(minutes) ?: return ""
    return "${value / 60}:${(value % 60).toString().padStart(2, '0')}"
}

data class MonthlyTargetProgress(
    val targetMinutes: Int,
    val actualMinutes: Int,
    val differenceMinutes: Int,
    val remainingMinutes: Int,
    val exceededMinutes: Int,
    val progressPercent: Float,
)

fun getValidMonthlyTargetMinutes(userData: UserData?): Int? =
    if (userData?.simpleMode == true) normalizeMonthlyTargetMinutes(userData.monthlyTargetMinutes) else null

fun calculateMonthlyTargetProgress(actualMinutes: Int, userData: UserData?): MonthlyTargetProgress? {
    val target = getValidMonthlyTargetMinutes(userData) ?: return null
    val actual = actualMinutes.coerceAtLeast(0)
    val difference = actual - target
    return MonthlyTargetProgress(
        targetMinutes = target,
        actualMinutes = actual,
        differenceMinutes = difference,
        remainingMinutes = (-difference).coerceAtLeast(0),
        exceededMinutes = difference.coerceAtLeast(0),
        progressPercent = ((actual.toFloat() / target) * 100f).coerceIn(0f, 100f),
    )
}
